### ISSUE TITLE
KeyError: 'assetId'

### ISSUE THREAD

**Issue by BadWindshield** [coursera-dl/coursera-dl#442]
When I tried to download the lectures for this Coursera class

https://www.coursera.org/learn/graph-analytics/home/welcome

the code crashed with the stack trace:

```
root[_get_open_course_asset_ids] Parsing JSON for asset_id <5b7TusBgEeWdnhIHQKZtcQ>.
Traceback (most recent call last):
  File "./coursera-dl", line 6, in <module>
    coursera_dl.main()
  File "/media/sf_Will/Personal/Docs/School/Coursera/coursera/coursera/coursera_dl.py", line 1056, in main
    if download_class(args, class_name):
  File "/media/sf_Will/Personal/Docs/School/Coursera/coursera/coursera/coursera_dl.py", line 1034, in download_class
    return download_on_demand_class(args, class_name)
  File "/media/sf_Will/Personal/Docs/School/Coursera/coursera/coursera/coursera_dl.py", line 987, in download_on_demand_class
    args.video_resolution)
  File "/media/sf_Will/Personal/Docs/School/Coursera/coursera/coursera/coursera_dl.py", line 331, in parse_on_demand_syllabus
    video_resolution, assets)
  File "/media/sf_Will/Personal/Docs/School/Coursera/coursera/coursera/api.py", line 73, in extract_links_from_lecture
    links, self._extract_links_from_lecture_assets(assets))
  File "/media/sf_Will/Personal/Docs/School/Coursera/coursera/coursera/api.py", line 127, in _extract_links_from_lecture_assets
    for open_course_asset_id in self._get_open_course_asset_ids(asset_id):
  File "/media/sf_Will/Personal/Docs/School/Coursera/coursera/coursera/api.py", line 153, in _get_open_course_asset_ids
    for element in dom['elements']]
KeyError: 'assetId'
```

The Python script works for some classes (e.g., cs-tech-interview) but not for others (e.g., regression-models).

**Comment by balta2ar** [comment]
It appears there elements can be of different type, e.g. asset type:
``` json
{"typeName": "asset", 
 "id": "3GpHlscoEeWZ9QqyXgsXow",
 "definition": {"name": "Преззентация к видео",
                "assetId": "grYbPMcoEeW7RxKvROGwrw"}}
```
and url:
``` json
{"id": "w1o7A6S-EeWVaQoOMF8a3w",
 "definition": {"name": "https://github.com/bcaffo/courses/raw/master/07_RegressionModels/pdfs/01_01.pdf",
                "url": "https://github.com/bcaffo/courses/raw/master/07_RegressionModels/pdfs/01_01.pdf"},
 "typeName": "url"}
```

I didn't expect that. Thank you for reporting, this needs to be fixed. As a quick hack, please replace this line https://github.com/coursera-dl/coursera-dl/blob/master/coursera/api.py#L153 as follows:
``` python
        return [element['definition']['assetId']
                for element in dom['elements'] if element['typeName'] == 'asset']
#                for element in dom['elements']]
```

**Comment by BadWindshield** [comment]
@balta2ar You suggested quick fix seems to work for the "graph-analytics" class.  Thanks very much!

However, I am experiencing another problem with the class "regression-models", which I will report in a separate ticket once I look into it.

**Comment by balta2ar** [comment]
@BadWindshield Good to know! Please report your issues. There have been recent changes related to downloading of supplementary files, and I couldn't possibly cover all the cases as Coursera platform API is not documented. Unfortunately, this is only trial and error. Thank you for your patience.

**Comment by rbrito** [comment]
@balta2ar, @BadWindshield, I just commited a slightly different fix with pull request #449, but I believe that @balta2ar's may be closer to "the truth".

Anyway, can anybody test that and tell me if things work with those changes?

@balta2ar, can you review my pull request and accept it or reject it, if you think that it is too dirty?

**Comment by iamashwin26** [comment]
@rbrito the problem with learning-how-to-learn seems to be corrected with the new fix

**Comment by balta2ar** [comment]
@rbrito I've prepared another pull request. I think it is more clear and general, than the solution that you have provided. Please take a moment to review the code.

For the record, I've tested new code on the following courses:
``` bash
coursera-dl -n --path . learning-how-to-learn \
    ml-foundations ml-regression vvedenie-mashinnoe-obuchenie \
    graph-analytics python-network-data regression-models \
    corporate-finance suanfa-jichu
```

Eventually, I managed to download almost everything, but there were problems along the way.

The first common problem is connection errors and 404 (426 sometimes). Sometimes our native downloader is able to get over it, sometimes I had to fall back to aria2c by adding this option: `--aria`.

Another problem is that with recent changes the downloaded tries to grab almost everything he can possibly reach. Sometimes it's downloading pages which I think are not necessary, e.g.:

``` js
"introduction",
[
    [
        "welcome-to-regression-models",
        {
            "u4f02fpu8cveioeaveh3pytdntarjbyzyhigjfodbxsywgvqb0rd-dhdgbcygy3zs-tn62p7m08wdsmxoq7n1vm-hxv5uf4gft-i8n_pbbvfqe4bjjnlqj5eek8cuxbhqc8pesyaeijrgmyf3lxvzhpq-ut2ipxkzgr1bxwtpzieja8wnnfsxkz46fhpt1cuqqxktkhcolibotm2j5derfgsluulj92-hui4hwyjt6hat5hyl5w6dzzv2aifotsbp_l9kcefdg51fpkyyxradyao4gbwdxrau1iwz1cbujan4mmtmsjfsg_2vkidpohjxzln_cbt2c9mfmvgzjoq7q": [
                [
                    "https://eventing.coursera.org/redirect/Kbp848YCgjxdpeGihUcxr4_PL8JM5tJxrEVr91e09B5KidCoGXWxe8c2AEKCoEXbBAaW4JMUUXI0R-KgvzNd7g.JywEThkg4eCE-iu2MYP7iw.u4F02Fpu8cveIoeaVeh3pyTDNTaRJByZyhIGJfoDbxSyWgvqb0RD-DHdGbcyGY3zs-TN62p7m08WdsmXoq7n1VM-HxV5Uf4gFt-i8N_pBBvfQE4bjJnLqJ5eeK8CuxbHQc8pesyaEIjrGmYf3lxvZHPq-ut2iPxKZgr1BXWTpzieJA8wNnfsXKZ46FHpt1cuQqXktkHcOliBOtM2J5derfGslUUlJ92-hUi4HWYjt6haT5hYL5w6DZZv2AifOtsBp_l9kceFDG51fpkyyXRAdyao4gbWdXraU1iwz1cbUjaN4MMtMsjFSg_2vkiDpohJxZLN_Cbt2c9mFMvgzjoq7Q",
                    "Kbp848YCgjxdpeGihUcxr4_PL8JM5tJxrEVr91e09B5KidCoGXWxe8c2AEKCoEXbBAaW4JMUUXI0R-KgvzNd7g.JywEThkg4eCE-iu2MYP7iw"
                ]
            ],
            "git": [
                [
                    "https://github.com/bcaffo/courses.git",
                    "courses"
                ]
            ]
        }
    ],
```

In this case it helps to restrict downloading file formats to some sane list as follows: `--formats 'mp4 txt pdf doc'`

**Comment by deepakjois** [comment]
I found that the following courses also have the same problems:

* erasmus-econometrics
* html-css-javascript-for-web-developers
* java-programming-design-principles

I can help test. Can anybody tell me exactly what I need to do. Should I rebase #451 on top of latest master and give it another shot?

**Comment by balta2ar** [comment]
@deepakjois Yes, please! Here is a guide just in case: https://help.github.com/articles/checking-out-pull-requests-locally/

**Comment by balta2ar** [comment]
I've just merged #451, please do `git pull` and try again. It's easier now, no need to follow the guide I have provided.

**Comment by deepakjois** [comment]
I can confirm that with the latest code, the three courses I listed above work fine.
